yaml validation

jjallaire opened the pull request.

cscheid committed:
minimal example of a validation error localized to source code

cscheid committed:
add fixme reminder

cscheid committed:
add a json schema EDSL, port config/types.ts to schema

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
add MappedString type and mappedString()

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
WIP: quarto on MappedString infra typechecks

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
recovers mapped strings from knitr via diff

cscheid committed:
compute sourceoffset from mappedstring

cscheid committed:
fix bug in sourceoffset

cscheid committed:
add minimal validation and error reporting for YAML front matter

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
add passably localized and pruned validation errors, start populating YAML schema

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
improve variable name

cscheid committed:
allow objectSchema to inherit; add allOfSchema; add integer and numeric schema

cscheid committed:
catch yaml parse error

cscheid committed:
fix bug on multiline cell yaml options

cscheid committed:
jupyter cell metadata

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
add WIP towards src/core/lib being a buildable independent library

cscheid committed:
small renaming refactor

cscheid committed:
add esbuild target stub

cscheid committed:
add binary search tests

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
WIP to autocompletion

cscheid committed:
add more WIP, now with schemas serialized and loaded in IDE

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
more WIP, minimal working completion demo

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
more WIP, autocomplete slightly more robust

cscheid committed:
add script to build JS and JSON assets

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
add more work in progress; new completion strategy based on line indentation

cscheid committed:
add per-completion suggest_on_accept

cscheid committed:
add per-completion suggest_on_accept

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
follow autocompletion into object and array completions

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
add more WIP, minimal working autocomplete demo inside IDE visual editor

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
autocompleter now works on qmd files and yml files, as well as on the visual editor

cscheid committed:
WIP: do not use catchall

cscheid committed:
merge

cscheid committed:
use cursor position to decide on key vs value context

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
remove unused import

cscheid committed:
refactoring completions and validation

cscheid committed:
WIP: working ajv in the browser, refactor all strings into mappedstrings

cscheid committed:
WIP: split yaml-schema into core-lib, use dependency injection setup

cscheid committed:
move promise into core-lib; allow error reporting to be parametrized in partition-cell-options to work on both deno and browser contexts

cscheid committed:
refactor yaml.src.js into ES6 automation.js module to be built with esbuild

cscheid committed:
remove debugger stmt

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
remove unnecessary make target

cscheid committed:
WIP: minimal YAML validation error reporting inside IDE

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
report multiple errors successfully, fix off-by-one column problem, localize invalidProperty errors

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
simple things working on qmd files

cscheid committed:
fix bad commit

cscheid committed:
use more sophisticated error narrowing heuristics

cscheid committed:
tweak narrowing rule for oneOf

cscheid committed:
more consistent behavior between autocomplete and validation

cscheid committed:
stamp out a few more mappedString bugs

cscheid committed:
pre-merge commit

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
be more defensive in the presence of errors in library results

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
add some docs

cscheid committed:
refactor some of the repeated calls

cscheid committed:
add deno fmt+lint pass

cscheid committed:
deno lint+fmt

cscheid committed:
facilitate deno lint running by ignoring core-lib.js

cscheid committed:
build assets after deno lint

cscheid committed:
don't end schema import names with underscores

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
add readonly annotations

cscheid committed:
remove unused methods

cscheid committed:
add format options to allow iife esbuild runs

cscheid committed:
fig.width -> fig.height etc, lint

cscheid committed:
convert ipynb to markdown only once

cscheid committed:
deno fmt+lint

cscheid committed:
deno lint+fmt

cscheid committed:
use module semantics since we can import in other js code

cscheid committed:
convert to an actual module that gets called in quarto build-js

cscheid committed:
use module rather than window global

cscheid committed:
new build artifact from build-js

cscheid committed:
update with new build-js cmd

cscheid committed:
fig.height -> fig-height, etc

cscheid committed:
add build-js command to quarto

cscheid committed:
build artifacts

cscheid committed:
use right version of core-lib from build artifacts

cscheid committed:
add build-js command

jjallaire reviewed (commented):
Looks great! Only a few small comments. It also look like some of the `FIXME` stuff needs resolving before merge, I'll leave it to you to decide which of those need to be done before we move on from the core. After we get this merged let's huddle up and decide how to tackle getting all the schemas written.

Also noted that we have one failing smoke test.

cscheid commented on src/core/esbuild.ts:
```diff
@@ -12,10 +12,14 @@ export async function esbuildCompile(
   args?: string[],
+  format?: "esm" | "iife",
 ): Promise<string | undefined> {
+  if (format === undefined) {
```
Can we go with `format = format ?? "esm"`? we've been bit with falsey issues before.

cscheid replied:
FYI I'm doing a pass that changes all of the "if (x === undefined) { x = ... }" I find in my source to `x = x ?? ...`.

jjallaire replied:
Would `format = format ?? "esm"` catch empty string? (I think we'd want to default empty string to "esm" so perhaps falsey is okay here? Otherwise I'm generally quite in favor of `??`

cscheid commented on package/scripts/common/quarto-deno:
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
```
Yes, it's on my TODO.

cscheid commented:
Here are some things I'm going to fix before merge:

- [ ] decide where the build artifacts go; add those locations to .gitignore
- [ ] move `src/resources/build/build-schema-file.ts` to a better place (you mentioned this already)
- [ ] clean scripts/remove them from repo
- [ ] add `buildJs` to `package/src/common/prepare-dist.ts`
- [ ] full pass over FIXMEs to grab low-hanging fruit
- [ ] handle failing smoke test

cscheid commented on src/core/lib/yaml-schema.ts:
```diff
@@ -0,0 +1,505 @@
+          Math.min(end.line + 1, nLines - 1),
+        );
+        for (const { lineNumber, content, rawLine } of lines) {
+          console.log(content);
```
Good point, thank you. I think I need a "plainPrint" or some similar function. IIRC, the issue I found was that using those calls added a lot of visual noise to the CLI error output, and it was hard to see the source in between all the "WARNING:" or "ERROR:" things being printed out.

cscheid commented on src/core/lib/break-quarto-md.ts:
```diff
@@ -0,0 +1,226 @@
+    cell_type: "markdown" | "code" | "raw" | "math",
+  ) => {
+    if (lineBuffer.length) {
+      // FIXME: understand why was this here. This makes our line
```
Ok great, thanks for the clarification.

cscheid commented on src/core/lib/README.md:
```diff
@@ -0,0 +1,19 @@
+# core/lib
```
I think it's fine to remove the README if building the artifacts is done automatically and cleanly like we're planning to do.

cscheid commented on src/command/build-js/cmd.ts:
```diff
@@ -0,0 +1,102 @@
+import { esbuildCompile } from "../../core/esbuild.ts";
+
+// FIXME this should be moved to a better place now that it's an actual TS module rather than a script
+import { buildSchemaFile } from "../../resources/build/build-schema-file.ts";
```
👍

cscheid committed:
rephrase undefined checks

cscheid committed:
remove js build artifacts and old js build scripts

cscheid committed:
temporarily add back the yaml.js build artifact

cscheid committed:
move build-schema-file.ts

cscheid committed:
use path from getCompletions and getLint instead of hardcoding it

cscheid committed:
update unit tests for latest yamlschema API

cscheid committed:
don't use ?? in core-lib because it's ES6-only

cscheid committed:
address fixme

cscheid committed:
add slightly better front matter schema; support empty front matter; don't validate front matter when it doesn't exist

cscheid committed:
copy files to right location in prepare-dist; use right imports; update build

cscheid committed:
hold off on prepare-dist building assets until I learn how to fix CI issues

cscheid commented:
@jjallaire this last TODO item [above](#issuecomment-970386637) has me stumped; https://github.com/quarto-dev/quarto-cli/runs/4229822735?check_suite_focus=true shows the failure of running esbuildCommand inside prepare-dist.

I'm checking with @dragonstyle to handle it. In the meantime, I've added the build assets back into the repo. `quarto build-js` still does the right thing. It's just that we need to remember to call it manually for now. This isn't great, but might not be a complete show-stopper: eg. the observable runtime still behaves this way in `main`.

dragonstyle committed:
Make Quarto Env Variables Available in `quarto-bld` environment

cscheid committed:
add back building of js assets

cscheid committed:
add autogen build assets

cscheid committed:
add more lenient codefold schema

cscheid committed:
echo raw html language chunks from ipynb

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
add validate document fn, deno fmt+lint

cscheid committed:
validate entire document instead of just front matter

cscheid committed:
add missing fix for mappedstring

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
do not patch locations

cscheid committed:
allow 'default' to be a valid html schema

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
validate only when validate-yaml: false is not set

cscheid committed:
report line errors on correct coordinate system

cscheid committed:
use last key to navigate, revert coordinate system changes to prioritize behavior on IDE

cscheid committed:
add tests on validation

cscheid committed:
minor refactoring to reuse utils in crossref across rest of test suite

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
aesthetics: add preamble comments

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
wip towards validation of all yaml in cli

cscheid committed:
ensure async calls on lib as well

cscheid committed:
fix mappedstring bug; add final async calls for breakQuartoMd/partitionCellOptions refactoring

cscheid committed:
validate project yaml, metadata, includes

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
schema for format language object

cscheid committed:
helper functions for reading validated yaml

cscheid committed:
use helper functions to read validated yaml, validate language format object

cscheid commented:
> I looked and it actually depends on the caller! From `resolveFormatsFromMetadata` it is as you said front matter, from `projectContext` it is _quarto.yml

Ok, we need to deal with this then, right? Let's flag this for later today. there's now a `readValidatedYamlFromFile` function that takes a schema and an error message; might be easier to just use that.

cscheid commented:
(Pulling the massive checklist from an inner comment to a top-level)

Here are some things I'm going to fix before merge:

- [x] decide where the build artifacts go; add those locations to .gitignore
  - for now, they're spread around a bit; but .gitignore shows them.
- [x] move `src/resources/build/build-schema-file.ts` to a better place
- [x] clean scripts/remove them from repo
- [x] add `buildJs` to `package/src/common/prepare-dist.ts`
- [x] full pass over FIXMEs to grab low-hanging fruit
- [x] handle failing smoke test
- [x] use "info" instead of needed console.log calls ~~create "plainPrint" or a similar printout function for our logging infra~~
- [x] remove stray console.log calls
- [x] use path given from `getCompletions` and `getLint` instead of hard-coding
- [x] resolve break-quarto-md differences between core and core-lib
  - [x] need discussion on partitionCellMapped's concerns (the difference is on validation and what library we use to parse YAML, which differs on IDE vs CLI) 
- [x] get a workable format schema
  - (resolution: defer full schema to a separate issue/feature. Still, schemas are such that tests pass now.)
- [x] validate the yaml chunks of the language blocks in the CLI
- [x] minimally improve break-quarto.md
  - (we'll move this to a separate feature, and the improvement is to add a "do not validate" yaml option to the chunk.)
- [x] allow user to force no YAML validation
- [ ] validate other sources of yaml
  - [x] project config
  - [x] metadata includes
  - [ ] language translations
  - [ ] reveal plugin yaml
  - [x] directory metadata
- [ ] aesthetics: check that we don't include from core/lib directly in cli.
  - [ ] meta: how do/can we add custom deno lint passes?
- [ ] add support for block_scalar in `buildAnnotated`
- [ ] context of error message should be shown in originalString, not mappedString.value. 
  - This matters in YAML multiple-block chunks where the error message might otherwise include discontinuous line ranges. 
  - How do we handle multiline-underlining in that case? We hope that it doesn't come up?

cscheid committed:
add schema parameter to includedMetadata

cscheid committed:
remove addressed TODO

cscheid committed:
add yaml validation

cscheid committed:
use new calling convention for withValidator

cscheid committed:
clean up api in validated-yaml.ts

cscheid committed:
fix API disparity in partitionCellOptionsMapped

cscheid committed:
fix syntax error

cscheid committed:
handle block_scalar strings

cscheid committed:
remove debugger statement

cscheid committed:
fix bug on locating yaml by path in autocompletion

cscheid committed:
Merge branch 'main' into feature/yaml-validation

cscheid committed:
deno fmt+lint

cscheid committed:
revert deno lint suggestions wrt require-await

cscheid committed:
rm debugger

cscheid committed:
add await

jjallaire committed:
remove spurious console.log

cscheid committed:
add another await

cscheid committed:
Merge branch 'feature/yaml-validation' of github.com:quarto-dev/quarto-cli into feature/yaml-validation

cscheid committed:
name plugin for withValidator caching

cscheid committed:
add revealjs as format string

cscheid committed:
make plugin format schema more lenient

cscheid committed:
path is optional

cscheid committed:
rm debugger

cscheid committed:
improve messages of validation errors at the top of the instance

cscheid committed:
report file path with when plugin configuration validation fails

jjallaire committed:
merge from origin/main

Merge remote-tracking branch 'origin/main' into feature/yaml-validation

# Conflicts:
#	src/command/render/render.ts
#	src/format/reveal/format-reveal-plugin.ts
#	src/project/project-context.ts

jjallaire committed:
Merge remote-tracking branch 'origin/main' into feature/yaml-validation

jjallaire committed:
regenerate ojs w/ fromEntries fix

cscheid committed:
actually polyfill fromEntries bc of observablehq runtime

cscheid committed:
use new patched version of observablehq's stdlib which does not call fromEntries
